File: DotDice/DiceParser.cs
using System.Runtime.CompilerServices;
using Pidgin;
using static Pidgin.Parser;

namespace DotDice
{
    public record ComparisonPoint(ComparisonOperator compOp, int value);

    public static class DiceParser
    {
        static Parser<char, T> Tok<T>(Parser<char, T> p)
            => Try(p).Before(SkipWhitespaces);

        // Parser for comparison operators
        public static readonly Parser<char, ComparisonOperator> comparisonOperator =
            Tok(
                Char('=')
                    .ThenReturn(ComparisonOperator.Equal)
                .Or(Char('>')
                    .ThenReturn(ComparisonOperator.GreaterThan))
                .Or(Char('<')
                    .ThenReturn(ComparisonOperator.LessThan))
            );

        // Parser for comparison point
        public static readonly Parser<char, ComparisonPoint> comparisonPoint =
            Tok(
                comparisonOperator
                .Then(
                    DecimalNum, (op, val) => new ComparisonPoint(op, val)
                )
            );

        // Parser for sort modifier
        public static readonly Parser<char, Modifier> sortModifier =
            Tok(
                String("sa")
                    .ThenReturn((Modifier)new SortModifier(SortDirection.Ascending))
                .Or(String("sd")
                    .ThenReturn((Modifier)new SortModifier(SortDirection.Descending)))
            );

        // Parser for success modifier
        public static readonly Parser<char, Modifier> successModifier =
            Tok(
                comparisonPoint.Map(cp => (Modifier) new SuccessModifier(cp.compOp, cp.value))
            );

        // Parser for failure modifier
        public static readonly Parser<char, Modifier> failureModifier =
            Tok(
                Char('f')
                .Then(
                    comparisonPoint,
                    (_, cp) => (Modifier) new FailureModifier(cp.compOp, cp.value)
                )
            );

        // Parser for explode modifier
        public static readonly Parser<char, Modifier> explodeModifier =
            Tok(
                Char('!').
                Then(
                    comparisonPoint,
                    (_, cp) => (Modifier) new ExplodeModifier(cp.compOp, cp.value)
                )
            );

        // Parser for compounding modifier
        public static readonly Parser<char, Modifier> compoundingModifier =
            Tok(
                Char('^').
                Then(
                    comparisonPoint,
                    (_, cp) => (Modifier) new CompoundingModifier(cp.compOp, cp.value)
                )
            );

        // Parser for reroll modifier
        public static readonly Parser<char, Modifier> rerollOnceModifier =
            Tok(
                String("ro")
                .Then(
                    comparisonPoint,
                    (_, cp) => (Modifier) new RerollOnceModifier(cp.compOp, cp.value)
                )
            );

        // Parser for reroll modifier
        public static readonly Parser<char, Modifier> rerollCompoundModifier =
            Tok(
                String("rc")
                .Then(
                    comparisonPoint,
                    (_, cp) => (Modifier) new RerollCompoundModifier(cp.compOp, cp.value)
                )
            );

        // Parser for keep modifier
        public static readonly Parser<char, Modifier> keepHighModifier =
            Tok(
                String("kh")
                .Then(
                    DecimalNum.Optional(),
                    (_,maybeVal) => 
                    {
                        var val = maybeVal.HasValue ? int.Abs(maybeVal.Value) : 1;
                        return (Modifier) new KeepModifier(val, true);
                    }
                )
            );

        // Parser for keep low modifier
        public static readonly Parser<char, Modifier> keepLowModifier =
            Tok(
               String("kl")
                .Then(
                    DecimalNum.Optional(),
                    (_,maybeVal) => 
                    {
                        var val = maybeVal.HasValue ? int.Abs(maybeVal.Value) : 1;
                        return (Modifier) new KeepModifier(val, false);
                    }
                )
            );

        // Parser for drop high modifier
        public static readonly Parser<char, Modifier> dropHighModifier =
            Tok(
                String("dh")
                .Then(
                    DecimalNum.Optional(),
                    (_,maybeVal) => 
                    {
                        var val = maybeVal.HasValue ? int.Abs(maybeVal.Value) : 1;
                        return (Modifier) new DropModifier(val, true);
                    }
                )
            );

        // Parser for drop low modifier
        public static readonly Parser<char, Modifier> dropLowModifier = 
            Tok(
                String("dl")
                .Then(
                    DecimalNum.Optional(),
                    (_,maybeVal) => 
                    {
                        var val = maybeVal.HasValue ? int.Abs(maybeVal.Value) : 1;
                        return (Modifier) new DropModifier(val, false);
                    }
                )
            );

        // Parser for a single modifier
        public static readonly Parser<char, Modifier> Modifier =
            successModifier
                .Or(failureModifier)
                .Or(dropHighModifier)
                .Or(dropLowModifier)
                .Or(keepHighModifier)
                .Or(keepLowModifier)
                .Or(rerollOnceModifier)
                .Or(rerollCompoundModifier)
                .Or(explodeModifier)
                .Or(compoundingModifier)
                .Or(sortModifier);

        // Parser for multiple modifiers
        public static readonly Parser<char, IEnumerable<Modifier>> Modifiers = Modifier.Many();

        // Parser for a constant
        public static readonly Parser<char, Roll> Constant =
            Tok(
                DecimalNum.Select(x => (Roll) new Constant(x))
            );

        // Parser for a die type section
        public static readonly Parser<char, DieType> dieTypeSection =
            Tok(
                Char('F')
                    .Map(_ => (DieType) new DieType.Fudge())
                .Or(Char('%')
                    .Map(_ => (DieType) new DieType.Percent()))
                .Or(DecimalNum
                    .Map(val => (DieType) new DieType.Basic(val)))
            );

        // Parser for a basic roll
        public static readonly Parser<char, Roll> basicRoll =
            Map(
                (maybeNum, _, die, mod) => 
                    (Roll) new BasicRoll(
                        maybeNum.HasValue ? maybeNum.Value : 1,
                        die,
                        mod
                    )
                ,
                DecimalNum.Optional(),
                Char('d'),
                dieTypeSection,
                Modifiers
            );

        // Top-level parser for any roll
        public static readonly Parser<char, Roll> Roll = basicRoll.Or(Constant);
    }
}

File: DotDiceF/Parser/DiceParser.fs
namespace DotDiceF.Parser

open DotDiceF.Core
//open DotDiceF.Core.ParserCombinators
open DotDiceF.Ast
open FParsec

/// The main parser for Roll20 dice roll expressions.
module DiceParser =
    /// Parser for comparison operators
    let pComparisonOperator : Parser<ComparisonOperator> =
        (pChar '=' >>. pReturn Equal)
        <|> (pChar '>' >>. pReturn GreaterThan)
        <|> (pChar '<' >>. pReturn LessThan)

    /// Parser for sort directions
    let pSortDirection : Parser<SortDirection> =
        (pChar 'a' >>. pReturn Ascending)
        <|> (pChar 'd' >>. pReturn Descending)

    /// Parser for sort modifier
    let pSortModifier : Parser<Modifier> =
        (pString "sa" >>. pReturn (Sort Ascending))
        <|> (pString "sd" >>. pReturn (Sort Descending))
        |> (fun p -> Sort <!> p)

    /// Parser for success modifier
    let pSuccessModifier : Parser<Modifier> =
        (fun op value -> Success (op, value)) <!> pComparisonOperator <*> pInteger

    /// Parser for failure modifier
    let pFailureModifier : Parser<Modifier> =
        (fun _ op value -> Failure (op, value)) <!> (pChar 'f') <*> pComparisonOperator <*> pInteger

    /// Parser for explode modifier
    let pExplodeModifier : Parser<Modifier> =
        let parseExclamation = pChar '!'
        let parseOperator = parseExclamation >>. (opt pComparisonOperator)
        let parseInt = parseOperator <*> pInteger

        (fun opOpt value ->
            let op = defaultArg opOpt Equal
            Explode (op, value)) <!> parseOperator <*> parseInt

    /// Parser for compounding modifier
    let pCompoundingModifier : Parser<Modifier> =
        let isCompounding = (fun _ op value -> Compounding (op, value)) <!> (pString "!!")  
        let possibleComp = (opt pComparisonOperator) 
        let int = pInteger
        (fun p value ->
            let op = Option.fold (fun _ x -> x) Equal p
            Compounding (op, value)) <!> possibleComp <*> int

    /// Parser for reroll modifier
    let pRerollModifier : Parser<Modifier> =
        (fun _ onlyOnce op value -> Reroll (op, value, onlyOnce))
            <!> (pChar 'r')
            <*> (opt (pChar 'o') |> (fun p -> Option.map (fun _ -> true) p |> Option.defaultValue false))
            <*> pComparisonOperator
            <*> pInteger

    /// Parser for keep modifier
    let pKeepModifier : Parser<Modifier> =
        let pKeepLowest =
            (fun _ count -> Keep (Option.defaultValue 1 count, false)) <!> (pString "kl") <*> pOptionalInteger
        let pKeepHighest =
            (fun _ count -> Keep (Option.defaultValue 1 count, true)) <!> (pString "kh") <*> pOptionalInteger
        pKeepHighest <|> pKeepLowest

    /// Parser for drop modifier
    let pDropModifier : Parser<Modifier> =
        let pDropLowest =
            (fun _ count -> Drop (Option.defaultValue 1 count, true)) <!> (pString "dl") <*> pOptionalInteger
        let pDropHighest =
            (fun _ count -> Drop (Option.defaultValue 1 count, false)) <!> (pString "dh") <*> pOptionalInteger
        pDropLowest <|> pDropHighest

    /// Parser for a single modifier
    let pModifier : Parser<Modifier> =
        pKeepModifier
        <|> pDropModifier
        <|> pRerollModifier
        <|> pExplodeModifier
        <|> pCompoundingModifier
        <|> pSuccessModifier
        <|> pFailureModifier
        <|> pSortModifier

    /// Parser for multiple modifiers
    let pModifiers : Parser<Modifier list> =
        many pModifier

    /// Parser for a constant
    let pConstant : Parser<Roll> =
        (fun value -> Constant value) <!> pInteger

    /// Parser for a die type section
    let pDieTypeSection : Parser<int * bool * bool> =
            (pChar 'F' >>. pReturn (0, true, false))
        <|> (pChar '%' >>. pReturn (0, false, true))
        <|> ((fun (dt: int) -> (dt, false, false)) <!> pInteger)

    /// Parser for a basic roll
    let pBasicRoll : Parser<Roll> =
        (fun numDice (dieType, isFudge, isPercentile) modifiers ->
            BasicRoll (Option.defaultValue 1 numDice, dieType, isFudge, isPercentile, modifiers))
        <!> pOptionalInteger
        <*> (pChar 'd' >>. pDieTypeSection)
        <*> pModifiers

    /// Parser for a roll group (either a basic roll or a constant)
    let pRollGroup : Parser<Roll> =
        pBasicRoll <|> pConstant

    /// Parser for a grouped roll
    let pGroupedRoll : Parser<Roll> =
        (fun rolls modifiers -> GroupedRoll (rolls, modifiers))
        <!> (pChar '{' >>. many1 (pRollGroup .>> opt (pChar ',')) .>> pChar '}')
        <*> pModifiers

    /// Top-level parser for any roll
    let pRoll : Parser<Roll> =
        pGroupedRoll <|> pBasicRoll